### ISSUE TITLE
interest from OpenPTV

### ISSUE THREAD

**Issue by alexlib** [pgodb/Peak-CNN#1]
Dear @pgodb - we'd be very interested in running your CNN with OpenPTV. Please let's discuss the best way to incorporate it. 
Alex

**Comment by pgodb** [comment]
Dear Alex, thanks for your interest. Unfortunately it will take a bit longer for me to upload code as the person approving these things is away on projects quite a lot at the moment. 

What I have prepared and plan to upload is a python implementation using tensorflow and numpy together with a little Jupyter notebook that implements a small test case. The function to apply the  peak detector accepts an image as numpy array and outputs the found peaks as numpy array so should be quite easy to integrate.

The training code besides the images needs 2d ground truth peak positions as input. If these are taken from 3d Tracking data or synthetic positions one would need to call some kind of „world2cam“ function from the respective camera model. Or, if the simpler approach using stacked peak detections is to be used one would need to call ones choice of peak detector on several images and add them together.

The code should be complete enough that it will be mostly a question of shoveling the right data through the interface, something that is probably best done by someone with expertise in the respective tracking codebase this is going to be integrated into.

Philipp

**Comment by pgodb** [comment]
It took longer than expected but I finally got the approval to publish the code. 

Best to start with the ipynb notebook. It shows basic training and application.

For incorporating this into an iterative approach like IPR you’ll likely want to include residual images during training. This is not included here but should be easy to implement with your existing tools. One just selects a percentage of particle training labels to still be found and then renders the others into a residual image. I typically use 100%, 50%, 10%, and 1% remaining
